Spark driver fails with K8s v1.6 out-of-box

Crashes with the following exception:

> Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://kubernetes.default.svc/api/v1/namespaces/default/pods/spark-dfsrw-1503358007989-driver. Message: Forbidden!Configured service account doesn't have access. Service account may have been revoked. User "system:serviceaccount:default:default" cannot get pods in the namespace "default"..
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:332)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:269)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:241)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:234)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleGet(OperationSupport.java:230)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handleGet(BaseOperation.java:745)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.get(BaseOperation.java:194)
	at org.apache.spark.scheduler.cluster.kubernetes.KubernetesClusterSchedulerBackend.liftedTree1$1(KubernetesClusterSchedulerBackend.scala:142)

It seems this is because the `default` service account does not have pod `edit` privilege any more under RBAC of K8s 1.6. Cluster admins are supposed to give only the barebone capability to the `default` account.

The recommendation is to add another service account, say `spark`, then add the necessary privileges:
```
$ kubectl create serviceaccount spark
$ kubectl create clusterrolebinding spark-edit --clusterrole edit --serviceaccount default:spark --namespace default
```

I did the above and I was hoping to use `--conf spark.kubernetes.authenticate.driver.serviceAccountName=spark`. But it seems the config key is no longer supported?

@foxish @mccheah

--- a/docs/running-on-kubernetes.md
+++ b/docs/running-on-kubernetes.md
@@ -114,9 +114,9 @@ the `default` service account used by the driver may not have appropriate pod `e
 for launching executor pods. We recommend to add another service account, say `spark`, with
 the necessary privilege. For example:
 
-  kubectl create serviceaccount spark
-  kubectl create clusterrolebinding spark-edit --clusterrole edit  \
-      --serviceaccount default:spark --namespace default
+    kubectl create serviceaccount spark
+    kubectl create clusterrolebinding spark-edit --clusterrole edit  \
+        --serviceaccount default:spark --namespace default
 
 With this, one can add `--conf spark.kubernetes.authenticate.driver.serviceAccountName=spark` to
 the spark-submit command line above to specify the service account to use.

--- a/docs/running-on-kubernetes.md
+++ b/docs/running-on-kubernetes.md
@@ -17,8 +17,10 @@ cluster, you may setup a test cluster on your local machine using
 * You must have appropriate permissions to create and list [pods](https://kubernetes.io/docs/user-guide/pods/),
 [ConfigMaps](https://kubernetes.io/docs/tasks/configure-pod-container/configmap/) and
 [secrets](https://kubernetes.io/docs/concepts/configuration/secret/) in your cluster. You can verify that
-you can list these resources by running `kubectl get pods` `kubectl get configmap`, and `kubectl get secrets` which
+you can list these resources by running `kubectl get pods`, `kubectl get configmap`, and `kubectl get secrets` which
 should give you a list of pods and configmaps (if any) respectively.
+  * The service account or credentials used by the driver pods must have appropriate permissions
+    as well for editing pod spec.
 * You must have a spark distribution with Kubernetes support. This may be obtained from the
 [release tarball](https://github.com/apache-spark-on-k8s/spark/releases) or by
 [building Spark with Kubernetes support](../resource-managers/kubernetes/README.md#building-spark-with-kubernetes-support).
@@ -106,6 +108,18 @@ the cluster.
 Finally, notice that in the above example we specify a jar with a specific URI with a scheme of `local://`. This URI is
 the location of the example jar that is already in the Docker image. Using dependencies that are on your machine's local
 disk is discussed below.
+
+When Kubernetes [RBAC](https://kubernetes.io/docs/admin/authorization/rbac/) is enabled,
+the `default` service account used by the driver may not have appropriate pod `edit` permissions
+for launching executor pods. We recommend to add another service account, say `spark`, with
+the necessary privilege. For example:
+
+  kubectl create serviceaccount spark
+  kubectl create clusterrolebinding spark-edit --clusterrole edit  \
+      --serviceaccount default:spark --namespace default
+
+With this, one can add `--conf spark.kubernetes.authenticate.driver.serviceAccountName=spark` to
+the spark-submit command line above to specify the service account to use.
 
 ## Dependency Management
 

--- a/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/DriverKubernetesCredentialsStep.scala
+++ b/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/DriverKubernetesCredentialsStep.scala
@@ -44,6 +44,7 @@ private[spark] class DriverKubernetesCredentialsStep(
       s"$APISERVER_AUTH_DRIVER_MOUNTED_CONF_PREFIX.$CLIENT_CERT_FILE_CONF_SUFFIX")
   private val maybeMountedCaCertFile = submissionSparkConf.getOption(
       s"$APISERVER_AUTH_DRIVER_MOUNTED_CONF_PREFIX.$CA_CERT_FILE_CONF_SUFFIX")
+  private val driverServiceAccount = submissionSparkConf.get(KUBERNETES_SERVICE_ACCOUNT_NAME)
 
   override def configureDriver(driverSpec: KubernetesDriverSpec): KubernetesDriverSpec = {
     val driverSparkConf = driverSpec.driverSparkConf.clone()
@@ -81,7 +82,17 @@ private[spark] class DriverKubernetesCredentialsStep(
             .endVolume()
           .endSpec()
         .build()
-    }.getOrElse(driverSpec.driverPod)
+    }.getOrElse(
+      driverServiceAccount.map {
+        account =>
+          new PodBuilder(driverSpec.driverPod)
+            .editOrNewSpec()
+            .withServiceAccount(account)
+            .withServiceAccountName(account)
+            .endSpec()
+            .build()
+      }.getOrElse(driverSpec.driverPod)
+    )
     val driverContainerWithMountedSecretVolume = kubernetesCredentialsSecret.map { secret =>
       new ContainerBuilder(driverSpec.driverContainer)
         .addNewVolumeMount()

--- a/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/DriverKubernetesCredentialsStep.scala
+++ b/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/DriverKubernetesCredentialsStep.scala
@@ -83,13 +83,12 @@ private[spark] class DriverKubernetesCredentialsStep(
           .endSpec()
         .build()
     }.getOrElse(
-      driverServiceAccount.map {
-        account =>
+      driverServiceAccount.map { account =>
           new PodBuilder(driverSpec.driverPod)
             .editOrNewSpec()
-            .withServiceAccount(account)
-            .withServiceAccountName(account)
-            .endSpec()
+              .withServiceAccount(account)
+              .withServiceAccountName(account)
+              .endSpec()
             .build()
       }.getOrElse(driverSpec.driverPod)
     )

